remote_file_system/server.py

```diff
@@ -140,33 +185,60 @@ def _dispatch_message(self, message: Message, client_ip_address: IPv4Address, cl
+                        # TODO send update notification below only if current time is before monitoring_expiration_timestamp
+                        # I am not sure about the delays between each subscribed client, so just put a curr time
+
+                        # TODO: I am not sure if at most one invocation semantics applies here. I will skip the implementation of it for now.
+
+                        if subscribed_client.monitoring_expiration_timestamp > curr_time:
+                            update_notification = UpdateNotification(
+                                file_name=message.file_name,
+                                content=message.content,
+                                modification_timestamp=modification_timestamp,
+                            )
+                            send_message(
+                                message=update_notification,
+                                recipient_ip_address=subscribed_client.ip_address,
+                                recipient_port_number=subscribed_client.port_number,
+                                max_attempts_to_send_message=1,
+                                timeout_in_seconds=5,
+                            )
+
+    def _check_for_duplicate_request_message(self, request_message: Message) -> bool:
+        if request_message.request_id in self.message_history:
+            return True
+        return False
+
+    def _add_message_to_history(self, request_id: uuid4, response_message: Message) -> None:
+        self.message_history[request_id] = response_message
+
+    def _remove_message_from_history(self, request_id: uuid4) -> None:
+        del self.message_history[request_id]
 
-            if is_successful:
-                curr_time = int(time.time())
-                for subscribed_client in subscribed_clients:
-                    # TODO send update notification below only if current time is before monitoring_expiration_timestamp
-                    # I am not sure about the delays between each subscribed client, so just put a curr time
-                    if subscribed_client.monitoring_expiration_timestamp > curr_time:
-                        update_notification = UpdateNotification(
-                            file_name=message.file_name,
-                            content=message.content,
-                            modification_timestamp=modification_timestamp,
-                        )
-                        send_message(
-                            message=update_notification,
-                            recipient_ip_address=subscribed_client.ip_address,
-                            recipient_port_number=subscribed_client.port_number,
-                            max_attempts_to_send_message=1,
-                            timeout_in_seconds=5,
-                        )
+    def _get_message_from_history(self, request_id: uuid4) -> Message:
+        return self.message_history[request_id]
```

Comment:
I think we can consider just calling the dictionary methods directly, so we may not need these functions. Is `_remove_message_from_history` being used? If we decide to keep these methods, we should change the type annotation from `uuid4` to `UUID`, just like the above comment

Reply:
i think it will be cleaner with the functions.